libfwupd/fwupd-client.c

```diff
@@ -1309,323 +1887,533 @@ fwupd_client_install_fd (FwupdClient *self,
+		g_task_return_error (task, g_steal_pointer (&error));
+		return;
+	}
+
+	/* success */
+	g_task_return_boolean (task, TRUE);
+}
+
+static void
+fwupd_client_install_release_bytes_cb (GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GTask) task = G_TASK (user_data);
+
+	if (!fwupd_client_install_bytes_finish (FWUPD_CLIENT (source), res, &error)) {
+		g_task_return_error (task, g_steal_pointer (&error));
+		return;
+	}
+
+	/* success */
+	g_task_return_boolean (task, TRUE);
+}
+
+static void
+fwupd_client_install_release_download_cb (GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GTask) task = G_TASK (user_data);
+	FwupdClientInstallReleaseData *data = g_task_get_task_data (task);
+	GChecksumType checksum_type;
+	GCancellable *cancellable = g_task_get_cancellable (task);
+	const gchar *checksum_expected;
+	g_autofree gchar *checksum_actual = NULL;
+
+	blob = fwupd_client_download_bytes_finish (FWUPD_CLIENT (source), res, &error);
+	if (blob == NULL) {
+		g_task_return_error (task, g_steal_pointer (&error));
+		return;
+	}
+
+	/* verify checksum */
+	checksum_expected = fwupd_checksum_get_best (fwupd_release_get_checksums (data->release));
+	checksum_type = fwupd_checksum_guess_kind (checksum_expected);
+	checksum_actual = g_compute_checksum_for_bytes (checksum_type, blob);
+	if (g_strcmp0 (checksum_expected, checksum_actual) != 0) {
+		g_task_return_new_error (task,
+					 FWUPD_ERROR,
+					 FWUPD_ERROR_INVALID_FILE,
+					 "checksum invalid, expected %s got %s",
+					 checksum_expected, checksum_actual);
+		return;
+	}
+
+	/* if the device specifies ONLY_OFFLINE automatically set this flag */
+	if (fwupd_device_has_flag (data->device, FWUPD_DEVICE_FLAG_ONLY_OFFLINE))
+		data->install_flags |= FWUPD_INSTALL_FLAG_OFFLINE;
+	fwupd_client_install_bytes_async (FWUPD_CLIENT (source),
+					  fwupd_device_get_id (data->device), blob,
+					  data->install_flags, cancellable,
+					  fwupd_client_install_release_bytes_cb,
+					  g_steal_pointer (&task));
+}
+
+static void
+fwupd_client_install_release_remote_cb (GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	g_autofree gchar *fn = NULL;
+	g_autofree gchar *uri_str = NULL;
+	g_autoptr(FwupdRemote) remote = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GTask) task = G_TASK (user_data);
+	FwupdClientInstallReleaseData *data = g_task_get_task_data (task);
+	GCancellable *cancellable = g_task_get_cancellable (task);
+
+	/* if a remote-id was specified, the remote has to exist */
+	remote = fwupd_client_get_remote_by_id_finish (FWUPD_CLIENT (source), res, &error);
+	if (remote == NULL) {
+		g_task_return_error (task, g_steal_pointer (&error));
+		return;
+	}
+
+	/* local and directory remotes may have the firmware already */
+	if (fwupd_remote_get_kind (remote) == FWUPD_REMOTE_KIND_LOCAL /*&& uri == NULL*/) {
```

Comment:
didn't you lose the `uri == NULL` from the original function?

Reply:
Agh yes, thanks. Good catch! Late night coding. I've restored it back so it worked the same was a before.